internal/generator/vector/source/kubernetes_logs.go

```diff
@@ -36,42 +42,94 @@ rotate_wait_ms = 5000
+// collection given a list of namespaces and containers and return a string that
+// is in a form directly usable by a vector kubernetes_log config. The result is
+// a set of file paths assumed to be at the well known location and structure of
+// CRIO pod logs.
+// The format rules:
+//
+//	namespaces:
+//	  namespace:     /var/log/pods/namespace_*/*/*.log
+//	  **namespace:   /var/log/pods/*namespace_*/*/*.log
+//	  **name*pace**: /var/log/pods/*name*pace*/*/*.log
+//	  namespace**:   /var/log/pods/namespace*/*/*.log
+//	containers:
+//	  container:    /var/log/pods/*/container/*.log
+//	  *cont**iner*:    /var/log/pods/*/*cont*iner*/*.log
+//	  cont**iner*:    /var/log/pods/*/cont*iner*/*.log
+func ContainerPathGlobFrom(namespaces, containers []string, extensions ...string) string {
+	paths := []string{}
+	for _, n := range namespaces {
+		paths = append(paths, fmt.Sprintf(crioNamespacePathFmt, normalizeNamespace(n)))
+	}
+	for _, c := range containers {
+		paths = append(paths, fmt.Sprintf(crioContainerPathFmt, removeDuplicateWildcards(c)))
+	}
+	for _, e := range extensions {
+		paths = append(paths, fmt.Sprintf(crioPathExtFmt, removeDuplicateWildcards(e)))
+	}
+	if len(paths) == 0 {
+		return ""
+	}
+	return joinContainerPathsForVector(paths)
 }
 
-func LokiLogStoreLogsPath(namespace string) string {
-	return fmt.Sprintf("/var/log/pods/%s_*/%%s*/*.log", namespace)
+func joinContainerPathsForVector(paths []string) string {
+	return fmt.Sprintf("[%s]", strings.Join(paths, ", "))
 }
 
-func VisualizationLogsPath(namespace string) string {
-	return fmt.Sprintf("/var/log/pods/%s_%%s-*/*/*.log", namespace)
+func normalizeNamespace(ns string) string {
+	if !strings.Contains(ns, "*") {
+		return fmt.Sprintf("%s_*", ns)
+	}
+	return fmt.Sprintf("%s_*", removeDuplicateWildcards(ns))
 }
 
-func ExcludeContainerPaths(namespace string) string {
-	return fmt.Sprintf("[%s]", strings.Join(
-		[]string{
-			fmt.Sprintf("%q", fmt.Sprintf(LogFilesMetricExporterLogsPath(namespace), constants.LogfilesmetricexporterName)),
-			fmt.Sprintf("%q", fmt.Sprintf(ElasticSearchLogStoreLogsPath(namespace), constants.ElasticsearchName)),
-			fmt.Sprintf("%q", fmt.Sprintf(LokiLogStoreLogsPath(namespace), constants.LokiName)),
-			fmt.Sprintf("%q", fmt.Sprintf(VisualizationLogsPath(namespace), constants.KibanaName)),
-			fmt.Sprintf("%q", fmt.Sprintf("/var/log/pods/%s_*/%s/*.log", namespace, "gateway")),
-			fmt.Sprintf("%q", fmt.Sprintf("/var/log/pods/%s_*/%s/*.log", namespace, "opa")),
-			fmt.Sprintf("%q", "/var/log/pods/*/*/*.gz"),
-			fmt.Sprintf("%q", "/var/log/pods/*/*/*.tmp"),
-		},
-		", ",
-	))
+func removeDuplicateWildcards(entry string) string {
+	result := []byte{}
+	var previous byte
+	for i, _ := range entry {
+		current := entry[i]
+		if current == '*' && current == previous {
+		} else {
+			result = append(result, current)
+		}
+		previous = current
+	}
+	return string(result)
```

Comment:
A cleaner way perhaps:
```suggestion
func removeDuplicateWildcards(entry string) string {
	for strings.Contains(entry, "**") {
		entry = strings.ReplaceAll(entry, "**", "*")
	}
	return string(entry)
```

Reply:
Or cleaner yet
```
func removeDuplicateWildcards(entry string) string {
    return strings.ReplaceAll(entry, "**", "*")
}
```

Reply:
What if I have more then 2 '*'?  I don't really like spinning through it byte by byte but this is the only way I could think to ensure multiples are addressed.  Separately also thinking there is a validation to ensure it is a valid NS

Reply:
> What if I have more then 2 '*'? I don't really like spinning through it byte by byte but this is the only way I could think to ensure multiples are addressed. Separately also thinking there is a validation to ensure it is a valid NS

In that case
```
var consecutiveWildcards = regexp.MustCompile(`\*+`)
...
func collapseWildcards(entry string) string {
	return consecutiveWildcards.ReplaceAllString(entry, "*")
}
```

Reply:
Or we can rewrite the original to be explicit that it works with runes and not bytes
```
func collapseWildcards(entry string) string {
	var result strings.Builder
	var previous rune
	result.Grow(len(entry))
	for _, current := range entry {
		if current == '*' && previous == '*' {
		} else {
			result.WriteRune(current)
		}
		previous = current
	}
	return result.String()
}
```
https://go.dev/play/p/5b_KVnV0gPM

Reply:
Replace all works, then a simple for loop in case of more than 2 consecutive.    Explicit and efficient